### PR TITLE
Cfy 3224 support rabbitmq security

### PR DESCRIPTION
**Pull request by geokala**

### CODE REVIEW

File: cloudify_agent/api/pm/base.py [hunk 1]
```diff
@@ -300,6 +329,37 @@ def validate_optional(self):
         self._validate_autoscale()
         self._validate_host()
 
+    def _get_broker_port(self):
+        """
+        Determines the broker port if it has not been provided. Only intended
+        to be called before self.broker_port has been set and after
+        self.broker_ssl_cert has been set.
+        """
+        if self.broker_ssl_cert == '':
```
Comment (nir0s) [inline cloudify_agent/api/pm/base.py]: if not self.broker_ssl_cert

Comment (nir0s) [inline cloudify_agent/api/pm/base.py]: might be nicer to add the ports to https://github.com/cloudify-cosmo/cloudify-agent/blob/master/cloudify_agent/api/defaults.py: so that they're available for both SSL and non-SSL..
and then do something like this:

```python
return defaults.BROKER_SSL_PORT if self.broker_ssl_cert else defaults.BROKER_PORT
```

File: cloudify_agent/api/pm/base.py [hunk 2]
```diff
@@ -274,6 +287,22 @@ def __init__(self, logger=None, **params):
                               backend=self.broker_url)
         self._celery.conf.update(
             CELERY_TASK_RESULT_EXPIRES=defaults.CELERY_TASK_RESULT_EXPIRES)
+        if self._get_ssl_cert_path() != '':
```
Comment (nir0s) [inline cloudify_agent/api/pm/base.py]: if not self._get_ssl_cert_path()

File: cloudify_agent/api/pm/base.py [hunk 3]
```diff
@@ -274,6 +287,22 @@ def __init__(self, logger=None, **params):
...
+            self._celery.conf.BROKER_USE_SSL = {
+                'ca_certs': self._get_ssl_cert_path(),
+                'cert_reqs': ssl.CERT_REQUIRED,
+            }
+
+    def _create_celery_conf(self):
+        utils.render_template_to_file(
+            template_path='pm/shared/worker_conf.py.template',
+            file_path=os.path.join(self.virtualenv,
```
Comment (nir0s) [inline cloudify_agent/api/pm/base.py]: we should discuss this. seems very fragile.

File: cloudify_agent/api/pm/base.py [hunk 4]
```diff
@@ -300,6 +329,37 @@ def validate_optional(self):
...
+        if self.broker_ssl_cert == '':
+            return 5672
+        else:
+            return 5671
+
+    def _create_ssl_cert(self):
+        """
+        Put the broker SSL cert into a file for AMQP clients to use.
+        """
+        if self.broker_ssl_cert != '':
```
Comment (nir0s) [inline cloudify_agent/api/pm/base.py]: if not self.broker_ssl_cert

File: cloudify_agent/resources/pm/nssm/nssm.conf.template [hunk 5]
```diff
@@ -13,6 +13,7 @@ echo Installing the agent as a windows service...
 --pidfile={{ pid_file }} ^
 --logfile={{ log_file }} ^
 --include={{ includes }} ^
+--config {{ work_dir }}/worker_conf.py
```
Comment (nir0s) [inline cloudify_agent/resources/pm/nssm/nssm.conf.template]: I think this will fail...
you need a caret at the end of the line...

Comment (geokala) [inline cloudify_agent/resources/pm/nssm/nssm.conf.template]: Correction pushed.

File: cloudify_agent/api/pm/base.py (hunk 3 above)
Comment (geokala) [inline cloudify_agent/api/pm/base.py]: Agreed- at best it's a bit ugly, and at worst it could fail interestingly. The problem with both this and the dynamic creation of the cert file are that I can't currently see a safer, saner, and (importantly) working alternative.

File: cloudify_agent/api/pm/base.py (hunk 1 above)
Comment (noakup) [inline cloudify_agent/api/pm/base.py]: :+1:

File: cloudify_agent/api/pm/base.py [hunk 6]
```diff
@@ -205,10 +208,16 @@ def __init__(self, logger=None, **params):
...
+        self.broker_ssl_cert = params.get(
+            'broker_ssl_cert', '')
         self.broker_ip = params.get(
             'broker_ip') or self.manager_ip
+        # Port must be determined after SSL cert has been set in order for
+        # intelligent default to work correctly
         self.broker_port = params.get(
-            'broker_port') or defaults.BROKER_PORT
+            'broker_port') or self._get_broker_port()
+        self.broker_user = params.get('broker_user', 'testuser')
```
Comment (noakup) [inline cloudify_agent/api/pm/base.py]: ?

File: cloudify_agent/api/pm/base.py [hunk 7]
```diff
@@ -217,10 +226,14 @@ def __init__(self, logger=None, **params):
             'name') or self._get_name_from_manager()
         self.queue = params.get(
             'queue') or self._get_queue_from_manager()
-        self.broker_url = params.get(
-            'broker_url') or defaults.BROKER_URL.format(
-            self.broker_ip,
-            self.broker_port)
+        # This is not retrieved by param as an option any more as it can make
+        # the _delete_amqp_queues function behave badly
```
Comment (noakup) [inline cloudify_agent/api/pm/base.py]: what do you mean? (regarding the comment)

File: cloudify_agent/api/pm/base.py (hunk 3 above)
Comment (noakup) [inline cloudify_agent/api/pm/base.py]: I didn't understand what's happening here. let's discuss tomorrow.

File: cloudify_agent/api/pm/base.py [hunk 8]
```diff
@@ -300,6 +329,37 @@ def validate_optional(self):
...
+            )
+
+    def _get_ssl_cert_path(self):
+        """
+        Determine what path the broker SSL cert should reside in.
+        """
+        if self.broker_ssl_cert == '':
+            return ''
+        else:
+            return os.path.join(self.workdir, 'broker.crt')
```
Comment (noakup) [inline cloudify_agent/api/pm/base.py]: I think this should be self._get_ssl_cert_path() and not 'broker.crt'

File: dev-requirements.txt [hunk 9]
```diff
@@ -1,11 +1,11 @@
-https://github.com/cloudify-cosmo/cloudify-dsl-parser/archive/master.zip
-https://github.com/cloudify-cosmo/cloudify-rest-client/archive/master.zip
-https://github.com/cloudify-cosmo/cloudify-plugins-common/archive/master.zip
-https://github.com/cloudify-cosmo/cloudify-script-plugin/archive/master.zip
+https://github.com/cloudify-cosmo/cloudify-dsl-parser/archive/3.3m5.zip
```
Comment (noakup) [inline dev-requirements.txt]: I am guessing you're changing that back to master..?

File: cloudify_agent/api/pm/base.py (hunk 7 above)
Comment (geokala) [inline cloudify_agent/api/pm/base.py]: If you pass the pre-created broker_url to _delete_amqp_queues, it's used by something which, at the back-end, chokes if passed something that isn't a DNS name/IP without port (you have to pass the port in separately).

Since accepting the pre-formatted URL causes breakages in situations where a username, password, and/or port are supplied, I decided it was more sensible to disallow it rather than to support tokenising it.

File: cloudify_agent/api/pm/base.py (hunk 3 above)
Comment (geokala) [inline cloudify_agent/api/pm/base.py]: It's creating the celery worker_conf.py, which is used to configure the SSL and the URL. The URL is easy to set outside of this (though if you see the comment in the template you'll note that the way we were doing it makes the credentials appear in a process listing). The SSL is a pain to set, as you really want to get the CERT_REQUIRED value in there, and trying to set it as a multiline variable in bash is doable but gets awkward fast.

As for moving it earlier in the lifecycle: I'm not sure we can- this is part of the agent install component anyway. When we're on the manager, this is easier to cope with as we can know where we will be putting agents. As soon as we start trying to deploy onto external nodes I don't think there's a better place for generating the required configs than this process manager module. The subclassed versions actually dropped some config in already (and indeed they had this call added to them).

File: cloudify_agent/api/pm/base.py (hunk 8 above)
Comment (geokala) [inline cloudify_agent/api/pm/base.py]: This is self._get_ssl_cert_path... not sure if I'm understanding your comment?

File: dev-requirements.txt (hunk 9 above)
Comment (geokala) [inline dev-requirements.txt]: Without meaning to sound like a Dr Who episode: Yes, the master is coming.

File: cloudify_agent/api/pm/base.py (hunk 1 above)
Comment (nir0s) [inline cloudify_agent/api/pm/base.py]: In general, anywhere we define the SSL or non-SSL ports, if a constants/defaults file exists, we can set them there, just like in the example above.

File: cloudify_agent/api/pm/base.py (hunk 3 above)
Comment (nir0s) [inline cloudify_agent/api/pm/base.py]: We should probably have worker_conf.py as a static file within cloudify-plugins-common so that it is accessible to the rest service as well.
We can then generate the configuration file (yaml) for it which is like cloudify_broker_conf.properties and read it from within worker_conf.py. To set and get the location for the generated properties file we can use pkg_resources.

File: cloudify_agent/api/pm/base.py [hunk 10]
```diff
@@ -205,10 +208,16 @@ def __init__(self, logger=None, **params):
...
         self.user = params.get('user') or getpass.getuser()
+        self.broker_ssl_cert = params.get(
+            'broker_ssl_cert', '')
         self.broker_ip = params.get(
             'broker_ip') or self.manager_ip
+        # Port must be determined after SSL cert has been set in order for
+        # intelligent default to work correctly
         self.broker_port = params.get(
-            'broker_port') or defaults.BROKER_PORT
+            'broker_port') or self._get_broker_port()
```
Comment (nir0s) [inline cloudify_agent/api/pm/base.py]:
```python
self.broker_port = params.get('broker_port', self._get_broker_port())
```

File: cloudify_agent/api/pm/base.py [hunk 11]
```diff
@@ -217,10 +226,14 @@ def __init__(self, logger=None, **params):
             'name') or self._get_name_from_manager()
         self.queue = params.get(
             'queue') or self._get_queue_from_manager()
-        self.broker_url = params.get(
```
Comment (nir0s) [inline cloudify_agent/api/pm/base.py]: @dankilman, do we actually need broker_url?

Comment (geokala) [inline cloudify_agent/api/pm/base.py]: I'm not certain. It may be used for something within the class or its descendants. Either way, as it had it before I wanted to ensure it was still present, but that it wouldn't cause problems by effectively overriding the supplied username, password, etc, which would then still be used by _delete_amqp_queues (see lines below, comment for which will be updated when I reach that stage).

**Comment by dankilman** [comment]
travis is failing

File: cloudify_agent/api/pm/base.py [hunk 12]
```diff
@@ -15,6 +15,8 @@
 
 import getpass
 import os
+from pkg_resources import resource_filename
```
Comment (dankilman) [inline cloudify_agent/api/pm/base.py]: `pkg_resources` is 3rd party, it should come after the `from celery import Celery` import

File: cloudify_agent/api/pm/base.py [hunk 13]
```diff
@@ -85,22 +91,23 @@ class Daemon(object):
         the ip address of the broker to connect to.
         defaults to the manager_ip value.
 
-    ``broker_port``
+    ``broker_ssl_enabled``:
```
Comment (dankilman) [inline cloudify_agent/api/pm/base.py]: Why did you remove the `broker_port` doc?

File: cloudify_agent/api/pm/base.py [hunk 14]
```diff
@@ -274,6 +295,28 @@ def __init__(self, logger=None, **params):
                               backend=self.broker_url)
         self._celery.conf.update(
             CELERY_TASK_RESULT_EXPIRES=defaults.CELERY_TASK_RESULT_EXPIRES)
+        if self.broker_ssl_enabled:
+            self._celery.conf.BROKER_USE_SSL = {
+                'ca_certs': self._get_ssl_cert_path(),
+                'cert_reqs': ssl.CERT_REQUIRED,
+            }
+
+    def _get_celery_conf_path(self):
```
Comment (dankilman) [inline cloudify_agent/api/pm/base.py]: as mentioned somewhere else, the location should probably be some place let intrusive than the site-packages dir that cloudify-agent lives in

File: cloudify_agent/resources/pm/detach/detach.template [hunk 15]
```diff
@@ -17,4 +17,5 @@
 --logfile={{ log_file }} \
 --include=$(cat {{ includes_path }}) \
 --detach \
--Ofair
\ No newline at end of file
+--config=cloudify.broker_conf \
```
Comment (dankilman) [inline cloudify_agent/resources/pm/detach/detach.template]: what is this?

File: cloudify_agent/resources/pm/initd/initd.conf.template [hunk 16]
```diff
@@ -30,7 +29,7 @@ CELERYD_LOG_LEVEL={{ log_level }}
 CELERYD_PID_FILE="{{ pid_file }}"
 CELERYD_LOG_FILE="{{ log_file }}"
 CELERYD_ENV_DIR={{ virtualenv_path }}
-CELERYD_OPTS="--events -Q {{ queue }} --hostname={{ name }} --autoscale={{ max_workers }},{{ min_workers }} --maxtasksperchild=10 -Ofair --without-gossip --without-mingle"
+CELERYD_OPTS="--events -Q {{ queue }} --hostname={{ name }} --autoscale={{ max_workers }},{{ min_workers }} --maxtasksperchild=10 -Ofair --without-gossip --without-mingle --config=cloudify.broker_conf"
```
Comment (dankilman) [inline cloudify_agent/resources/pm/initd/initd.conf.template]: what is this?

File: cloudify_agent/resources/pm/shared/broker.crt.template [hunk 17]
```diff
@@ -0,0 +1 @@
+{{ cert }}
```
Comment (dankilman) [inline cloudify_agent/resources/pm/shared/broker.crt.template]: if all you are doing is replacing a single value, then there really isn't any need of using a template in this case is there? simply write the content of `cert` into the file directly

File: cloudify_agent/resources/pm/initd/initd.conf.template (hunk 16 above)
Comment (geokala) [inline cloudify_agent/resources/pm/initd/initd.conf.template]: Celery broker configuration- specifically, this one reads from the broker_conf config file that you've commented on elsewhere.

This allows:
a) Broker SSL configuration in a sane manner (it can be exceedingly ugly to configure from exported vars).
b) Brokers that don't display the username+password (which reside in the URL) when anyone on the system runs ps.

Well, (b) would be true if I'd actually removed the hostname from this one. *making note*

**Comment by geokala** [comment]
Cannot pass travis until cloudify-plugins-common is merged.

Cannot pass appveyor as it cannot download get-pip?

File: cloudify_agent/resources/pm/initd/initd.template [hunk 18]
```diff
@@ -134,6 +134,9 @@ CELERYD_PID_DIR=`dirname $CELERYD_PID_FILE`
 # Extra start-stop-daemon options, like user/group.
 if [ -n "$CELERYD_CHDIR" ]; then
     DAEMON_OPTS="$DAEMON_OPTS --workdir=$CELERYD_CHDIR"
+
+    # Ensure we start in this directory if defined so that we can find our config
+    cd $CELERYD_CHDIR
```
Comment (dankilman) [inline cloudify_agent/resources/pm/initd/initd.template]: remove

File: cloudify_agent/api/pm/base.py [hunk 19]
```diff
@@ -781,6 +862,7 @@ def create_enable_cron_script(self):
             start_command=self.start_command(),
             status_command=self.status_command()
         )
+        self._logger.debug('TTT: {sc}'.format(sc=self.start_command()))
```
Comment (dankilman) [inline cloudify_agent/api/pm/base.py]: logging should probably be removed

File: cloudify_agent/shell/commands/daemons.py [hunk 20]
```diff
@@ -21,6 +21,8 @@
 from cloudify_agent.api.factory import DaemonFactory
 from cloudify_agent.shell import env
 from cloudify_agent.shell.decorators import handle_failures
+from cloudify.context import BootstrapContext
```
Comment (dankilman) [inline cloudify_agent/shell/commands/daemons.py]: these two new imports are cloudify related 3rd party so they should some between to currently two existing sections

File: cloudify_agent/tests/api/pm/test_base.py [hunk 21]
```diff
@@ -32,7 +32,9 @@ def setUp(self):
         self.daemon = Daemon(
             manager_ip='manager_ip',
             queue='queue',
-            name='name'
+            name='name',
+            broker_user='guest',
```
Comment (dankilman) [inline cloudify_agent/tests/api/pm/test_base.py]: no real need to pass this in all the tests here as this is the default value in base.py anyway

Comment (geokala) [inline cloudify_agent/tests/api/pm/test_base.py]: True, but this means that if someone changes that method's signature then this test will fail (rightly), so probably worth leaving in?

Comment (dankilman) [inline cloudify_agent/tests/api/pm/test_base.py]: cool

File: cloudify_agent/shell/commands/daemons.py [hunk 22]
```diff
@@ -133,6 +167,38 @@ def create(**params):
...
+            bootstrap_agent
+        )
+
+        try:
+            attributes['broker_user'] = broker_user
+            attributes['broker_pass'] = broker_pass
+            attributes['broker_ssl_enabled'] = \
+                bootstrap_agent.broker_ssl_enabled
+            attributes['broker_ssl_cert'] = bootstrap_agent.broker_ssl_cert
+        except AttributeError:
```
Comment (dankilman) [inline cloudify_agent/shell/commands/daemons.py]: is this really necessary? I mean, this code will always run as 3.3+ so an `AttributeError` will never be raised

Comment (geokala) [inline cloudify_agent/shell/commands/daemons.py]: It might help in the case of some local testing where the bootstrap agent isn't really a bootstrap agent but instead, e.g. an empty dict.

If that's not a concern here then it should be safe to remove.

Comment (dankilman) [inline cloudify_agent/shell/commands/daemons.py]: i think we should remove it
